capstone/src/test/java/com/google/sps/servlets/ChallengeServletTest.java

```diff
@@ -0,0 +1,168 @@
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private ChallengeServlet challengeServlet;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Add test data
+    ImmutableList.Builder<Entity> challenge = ImmutableList.builder();
+    challenge.add(createChallenge(CHALLENGE_NAME));
+    datastore.put(challenge.build());
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    challengeServlet = new ChallengeServlet();
+  }
+
+  private Entity createChallenge(String text) {
+    Entity challengeEntity = new Entity("Challenge");
+    challengeEntity.setProperty("name", text);
+    challengeEntity.setProperty("dueDate", this.getDueDate());
+    challengeEntity.setProperty("votes", new ArrayList<String>());
+    challengeEntity.setProperty("timestamp", System.currentTimeMillis());
+    return challengeEntity;
+  }
+
+  /**
+   * Sets due date to midnight, 7 days from when challenge is posted.
+   *
+   * @return due date in milliseconds
+   */
+  private long getDueDate() {
+    LocalDateTime now = LocalDateTime.now();
+    LocalDateTime dueDate = now.plusDays(7).withHour(23).withMinute(59).withSecond(59).withNano(0);
+    return Timestamp.valueOf(dueDate).getTime();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+    responseWriter = null;
+    datastore = null;
+    challengeServlet = null;
+  }
+
+  @Test
+  public void doGet_testChallengeName() throws IOException {
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    assertTrue(response.contains(CHALLENGE_NAME));
+  }
+
+  @Test
+  public void doGet_testUserNotCompleted() throws IOException {
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    assertTrue(response.contains("false"));
+  }
+
+  @Test
+  public void doGet_userNotLoggedIn() throws IOException {
+    helper.setEnvIsLoggedIn(false);
+
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+
+    assertTrue(response.contains("Oops an error happened!"));
+  }
+
+  @Test
+  public void doPost_userLoggedIn() throws IOException, EntityNotFoundException {
+    when(mockRequest.getParameter("name")).thenReturn(NEW_CHALLENGE);
+
+    challengeServlet.doPost(mockRequest, mockResponse);
+    // id of Challenge increments for each new Challenge that is added
+    Key challengeKey = KeyFactory.createKey("Challenge", CHALLENGE_ID);
```

Comment:
Hmm what do you mean by this? Is this a feature that you wrote in the ChallengeServlet class or a feature of the LocalTestService ?

Reply:
It's a feature of LocalTestService so if there is one thing in LocalTestService datastore it will have id of 1 and the following one that's added will have id of 2

Reply:
Ah, got it